### ISSUE TITLE
Documentation: API reference to do list

### ISSUE THREAD

**Issue by benanne** [Lasagne/Lasagne#203]
Here is a to do list for the API reference. Each of these could be a separate PR, or you can tackle multiple in a single PR. I guess it's a good idea to post here if you plan to work on one or more of these, so we don't end up doing any double work. Then I'll add your name on the list for an overview. Let's try to get this done within the next two weeks!

- [x] convert `lasagne.layers.helper` docs to numpydoc format (@MartinThoma)
- [x] convert `lasagne.layers.base` docs to numpydoc format (@MartinThoma)
- [x] convert `lasagne.layers.input` docs to numpydoc format (@MartinThoma)
- [x] complete `lasagne.layers.dense` docs and convert to numpydoc format (@skaae)
- [x] document `lasagne.layers.conv` (@benanne)
- [x] document `lasagne.layers.pool` (@ebenolson, @benanne)
- [x] document `lasagne.layers.noise` (@skaae)
- [x] complete `lasagne.layers.shape` docs and convert to numpydoc format (@skaae)
- [x] document `lasagne.layers.merge` (@skaae)
- [x] figure out how to make the docs for `corrmm`, `cuda_convnet` and `dnn` show up on readthedocs (these submodules cannot be imported without a gpu) (@benanne)
- [x] document `lasagne.layers.dnn` (@benanne)
- [x] document `lasagne.layers.corrmm` (@benanne)
- [x] document `lasagne.layers.cuda_convnet` (@benanne)
- [x] complete `lasagne.updates` docs and convert to numpydoc format (@skaae)
- [x] document `lasagne.init` (@skaae)
- [x] document `lasagne.nonlinearities` (@MartinThoma)
- [x] complete `lasagne.objectives` docs and convert to numpydoc format (@f0k)
- [x] complete `lasagne.regularization` docs and convert to numpydoc format (@skaae)
- [x] complete `lasagne.utils` docs and convert to numpydoc format (@JeffreyDF)
- [x] change the layout so it looks decent (@f0k)
- [x] get rid of string representations of objects in docstrings everywhere (e.g. `W=<lasagne.init.GlorotUniform object at 0x7f699026a690>`) (@benanne)
- [x] split `lasagne.layers` documentation into separate pages for each submodule (@f0k)

**Comment by MartinThoma** [comment]
I can do the easy stuff (to numpydoc conversions of `lasagne.layers.helper`, `lasagne.layers.base`, `lasagne.layers.input`) today evening

**Comment by benanne** [comment]
Awesome :) I'll try to write something up for the convolutional layers, since those have a ton of arguments. I guess I'll do the same for the other convolution implementations as well, once we figure out how to make them show up in the docs.

**Comment by MartinThoma** [comment]
I am not sure how familiar people are with numpydoc. If read about it the first time a couple of days ago and found [A Guide to NumPy/SciPy Documentation](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt) and [Example NumPy Style Python Docstrings](http://sphinxcontrib-napoleon.readthedocs.org/en/latest/example_numpy.html#example-numpy-style-python-docstrings) helpful.

**Comment by benanne** [comment]
That's great, thanks! I was planning to just copy your examples from `nonlinearities` but that will definitely be helpful as well :p

**Comment by skaae** [comment]
I can do updates. Is nonlinearities.py in the correct doc format?

**Comment by benanne** [comment]
Yep, it should be! There are a few oddities to work out (like the multiple references thing) but we can figure that out later.

**Comment by MartinThoma** [comment]
@skaae Yes, it is. See https://github.com/Lasagne/Lasagne/pull/204 (e.g. https://github.com/MartinThoma/Lasagne/blob/docs/lasagne/layers/base.py) for more examples of Numpydoc

**Comment by skaae** [comment]
I created a PR for updates doc: https://github.com/Lasagne/Lasagne/pull/211

I'll do noise aswell.

**Comment by skaae** [comment]
PR for noise.py https://github.com/Lasagne/Lasagne/pull/212

**Comment by skaae** [comment]
Starting on init.py. Can somebody point me to the He paper?  and is `c01b` related to convolutions?

**Comment by benanne** [comment]
On a roll :) Here's the He et al. paper: http://arxiv.org/abs/1502.01852
`c01b` is a notation used by pylearn2 to indicate the different axis order used by cuda-convnet. `c` is for 'channels', `b` is for 'batch' and `0` and `1` are the spatial dimensions. cuda-convnet has the batch dimension at the end (fastest varying dimension), hence `c01b`. The Theano default is sometimes referred to as `bc01`.

**Comment by skaae** [comment]
Yes, figured I better start contributing to getting the release done and documentation seems like an easy place to start :)

**Comment by JeffreyDF** [comment]
I'll try to do the utils one.

**Comment by skaae** [comment]
I have few question about the Glorot init.

  1.   c01b should be used when cuda-convnet kernels are used?
  2.   In the Glorot weights are sampled from 
    normal std = self.gain * np.sqrt(2.0 / ((n1 + n2) * receptive_field_size)). For Dense layer `receptive_field_size` is 1 and the equation then looks like eq 12 from http://machinelearning.wustl.edu/mlpapers/paper_files/AISTATS2010_GlorotB10.pdf. The ReLU gains makes sense because it is linear?

Why do the Glorot sample from the normal distribution? I would expect it to sample from the uniform distribution as in equation 16? Shouldn't the default gain for GlorotNormal and GlorotUniform be np.sqrt(6)/np.sqrt(2)  to make them equal to eq 16?

**Comment by JeffreyDF** [comment]
@skaae 

For 1: yes, cuda-convnet uses c01b.

**Comment by benanne** [comment]
>  1.   c01b should be used when cuda-convnet kernels are used?

the cuda-convnet kernels use c01b ordering by default. However, for compatibility reasons, `lasagne.layers.cuda_convnet.Conv2DCCLayer` adds the necessary dimshuffles so that bc01 ordering can be used, like elsewhere in the library. So then you can just use regular initializers. These additional dimshuffles can be disabled for performance reasons (by setting `dimshuffle=False` when creating the layer). In that case, you should use the c01b versions of the initializers as well.

So in the usual case, the answer is actually no, because Lasagne takes care of all the necessary dimshuffling.

>  2.   In the Glorot weights are sampled from 
    normal with  `std = self.gain * np.sqrt(2.0 / ((n1 + n2) * receptive_field_size))`. For Dense layer `receptive_field_size` is 1 and the equation then looks like eq 12 from http://machinelearning.wustl.edu/mlpapers/paper_files/AISTATS2010_GlorotB10.pdf. The ReLU gains makes sense because it is linear?

I'm not sure I understand the question. The `2.0` under the square root here comes from the fact that we're actually using the average of the fan-in and the fan-out (i.e. `(n1*receptive_field_size + n2*receptive_field_size) / 2.0`), so this is not the ReLU gain. The ReLU gain is contained in `self.gain`.

> Why do the Glorot sample from the normal distribution? I would expect it to sample from the uniform distribution as in equation 16?

You should use `GlorotNormal` if you want the normal distribution, `GlorotUniform` if you want the uniform version. Both will work, Glorot's proposed scaling mechanism is independent of the distribution you choose to use (provided that it is symmetric around 0). That is why the `Glorot` superclass exists, so that this mechanism can be combined with other distributions (which may be added in the future).

> Shouldn't the default gain for `GlorotNormal` and `GlorotUniform` be np.sqrt(6)/np.sqrt(2)  to make them equal to eq 16?

No, the `Normal` and `Uniform` initializers can be created using an `std`, and they take care of these additional scale factors themselves.

**Comment by skaae** [comment]
Question 2: I guess the question is that the code use: `std = self.gain * np.sqrt(2.0 / ((n1 + n2)` which i guess is equation 12 from the paper? But when you set gain to 'relu' then the gain is set to sqrt(2)?

Secondly I dont get if equation 16 from the paper should be used?

**Comment by benanne** [comment]
If you use `GlorotUniform`, what happens is equivalent to equation 16. It's just that the sqrt(6) in that equation is pulled apart into a sqrt(2) in `Glorot` and a sqrt(3) somewhere in `Uniform`.

**Comment by skaae** [comment]
Ok, that makes sense :)

**Comment by f0k** [comment]
Just throwing in that I've checked the equations when reviewing #180 and can confirm that Sander did everything correctly. If you want to check for yourself, lookup the standard deviation for continuous uniform distributions and use it to pull apart Equation 16 into a part computing the standard deviation (that's in `Glorot`) and a part computing the range of a uniform distribution matching this standard deviation (that's in `Uniform`).

**Comment by benanne** [comment]
I'm currently trying to figure out how to make the specialized convolution modules import. The problem is that they all raise an `ImportError` if `theano.config.device` does not start with `'gpu'`. Obviously this is not the case on the readthedocs.org server, so these modules never show up in the docs.

We need a way to 'fool' these modules so they become importable. It would be nice if we could do this without any modifications to the module code itself. I looked into mocking `theano.config.device`, but it's a property which cannot be assigned to, so that complicates matters.

If that's not possible, we'll just have to change this check so it makes an exception for readthedocs.org somehow.

If anybody has any ideas on how to solve this elegantly, please do share.

**Comment by ebenolson** [comment]
You can change the property like this:

    device_param = [p for p in theano.configparser._config_var_list if p.fullname == 'device'][0]
    device_param.allow_override = True
    theano.config.device = 'gpu'

Whether that is elegant (or functional) I'm not sure...

**Comment by dnouri** [comment]
> We need a way to 'fool' these modules so they become importable. It 
> would be nice if we could do this without any modifications to the 
> module code itself. I looked into mocking theano.config.device, but 
> it's a property which cannot be assigned to, so that complicates 
> matters.
> 
How about mocking theano.config instead?

> >>> impot theano
> >>> from unittest.mock import Mock
> >>> theano.config = Mock(device='gpu')
> >>> theano.config.device
> 'gpu'

**Comment by benanne** [comment]
doh... hadn't thought that far ahead :) That should work! Unfortunately I can only test it by pushing and waiting to see what happens (I have no machine without a GPU available to try it on).

**Comment by f0k** [comment]
> I have no machine without a GPU available to try it on

You could probably just remove nvcc from your `PATH` to try it (or anything else that breaks CUDA for Theano).

**Comment by ebenolson** [comment]
>  figure out how to prevent string representations of objects (.e.g. W=<lasagne.init.GlorotUniform object at 0x7f699026a690>) from appearing in function signatures and fix it everywhere

I looked into this a bit, seems like it's a [known issue](https://github.com/sphinx-doc/sphinx/issues/759).
Some workarounds here: http://stackoverflow.com/questions/7228797/sphinx-ext-autodoc-keeping-names-of-constants-in-signature

**Comment by benanne** [comment]
So basically having the signature as the first line of the docstring would be sufficient? Feel free to try that out if you want, otherwise I'll give it a go later and see what happens.

**Comment by ebenolson** [comment]
Just tried that, seems to work ok. Looks like it needs backslashes for line continuation to pick up the signature properly. 

Is this something you want to add only where needed, or standardize on for all docstrings? If the latter I may try writing a script to autogenerate it.

**Comment by benanne** [comment]
It's probably easier to do this only when necessary, I think.

**Comment by dnouri** [comment]
> Is this something you want to add only where needed, or standardize 
> on for all docstrings? If the latter I may try writing a script to 
> autogenerate it.
> 
How about making a nicer __repr__ for Glorot instead?

**Comment by benanne** [comment]
That was just one example though, the docs are littered with instances of this problem. In the long term implementing `__repr__` where appropriate is probably a good idea, but I don't think that covers everything (what about `nonlinearity=<function rectify at 0x7f6fa07812a8>` for example).

**Comment by ebenolson** [comment]
There's also uses of Theano objects (`T.mean` etc) which would need to be wrapped somehow.

**Comment by f0k** [comment]
Another alternative is to move the handling of default values into the constructor body. I.e., have `None` as the default in the signature, and replace it with, e.g., a `GlorotUniform()` instance in the body. However, this would break things like `nonlinearity=None` for identity, and `b=None` for biasless layers; we'd need to support `False` for that then. (I don't really think this would be better, just wanted to throw it into the discussion. I think it would even be regarded as more pythonic, given that all examples looking ugly in the documentation are actually modifiable objects that you might better want to create for each execution rather than once on definition of a method.)

**Comment by benanne** [comment]
I'm not a fan of having defaults handled in the body when it's not strictly necessary. I think it reduces readability a lot (in the sense that it becomes much harder to see what the code does at a glance).

I think all the examples of this in the docs are actually treated as immutable objects (even if they technically aren't) so I think it's okay to do it this way. Strictly speaking it's not 'nice' to create all these instances at loading time, but it doesn't really cause any noticeable overhead anyway. As usual I am erring on the side of pragmatism here :)

**Comment by benanne** [comment]
We could still use some help with `dense`, `shape` and `merge`, if anyone feels up to the task :) Don't bother with `objectives` for now (see #173).

I will do the remaining conv/pool implementation docs, and once #182 is merged we can also finally rebase/merge @MartinThoma's work. So all things considered I think we are almost there!

**Comment by skaae** [comment]
I'll do dense now and do shape + merge tomorrow if no one does it before then.

**Comment by f0k** [comment]
Regarding the style sheet: I played around with Sphinx a bit, and the "classic" Sphinx theme (as in [the Python documentation](https://docs.python.org/)) looks way better than the bootstrap theme we have now. The main issues with the current theme are:
* Font sizes vary wildly. E.g., on http://lasagne.readthedocs.org/en/latest/modules/layers.html, I find it difficult to see where a new class/function starts, the parameter descriptions use a larger font than everything else, and the heading for submodules is ridiculously large and bold red.
* "See also" boxes look like crap, with red and white font on purple background.
* The navigation is not very prominent. I strongly prefer permanently visible side bars over a drop-down menu when reading on a computer screen (i.e., enough horizontal space).

For "the parameter descriptions use a larger font", the reason is simply that numpydoc renders them as `<blockquote>` elements, and all themes on http://bootswatch.com/ use a larger font for that than for normal body text. I've tried using `sphinx.ext.napoleon` instead of numpydoc, but that doesn't support all of numpydoc (e.g., no class attributes) and generally looks worse (less vertical space, except for the type of return values, which get an extra line).

Soo... we'd either need to heavily modify the bootstrap theme and the bootswatch templates, or switch to a different theme. There are a bunch of those coming with Sphinx: http://sphinx-doc.org/theming.html
Was there any particular reason for choosing bootstrap, or can we just as well choose another one? How do you feel about a sticky side bar navigation?

**Comment by benanne** [comment]
No particular reason, I thought it looked nice. Now that we're doing numpydoc style docs... not so much :)

A nice thing about the bootstrap-based themes is that they look good on any device. Then again, I guess not too many people browse docs on their phones / tablets, so maybe that's not too important. 

My plan was to look at how these guys made it look good (they have a custom stylesheet): http://scikit-bio.org/docs/latest/index.html
But I'm okay with changing the theme to something that looks nice by default, if such a thing exists. Saves us a bunch of trouble.

**Comment by dnouri** [comment]
+1 to getting rid of the bootstrap style.  I think the [readthedocs theme](http://docs.readthedocs.org/en/latest/theme.html) looks really good (also on every device).  I agree that the classic theme also looks way better.  Not sure how well they work with numpydoc, but it can't be much worse than now.

**Comment by benanne** [comment]
Ha, looking at the objective docs (which are still in the old format and will probably be rewritten soon), I now remember why I decided to look for a different theme ;) http://lasagne.readthedocs.org/en/latest/modules/objectives.html

With numpydoc it does indeed look much better than the alternatives, though.

**Comment by f0k** [comment]
> I now remember why I decided to look for a different theme

Uh, you're right :) There's still a lot one could improve, and also we're now displaying an ad in the navigation bar, but it's a better starting point than the bootstrap theme. Things that bother me:
* "Parameters:" and "Returns:" is typeset quite a few pixels too low
* the "src" links are missing (when I compile locally, you can flip between documentation and source code for each function/class)
* there is an "Edit on GitHub" link that just links to the `.rst` file, it would be nice to have that for docstrings as well
* `lasagne.layers` is rendered as a single, much too long page. We should have an overview table ([as in numpy](http://docs.scipy.org/doc/numpy/reference/routines.fft.html)) and separate pages for the submodules.

**Comment by benanne** [comment]
Splitting `lasagne.layers` was already on the todo list :) I agree wit the other things as well.

What ad are you talking about though?

**Comment by f0k** [comment]
> What ad are you talking about though?

The one on the bottom left on http://lasagne.readthedocs.org. It's explained here: http://docs.readthedocs.org/en/latest/sponsors.html#sponsorship-information

**Comment by benanne** [comment]
Weird, I don't see it. I thought it'd be because of adblock, but I don't see it when I disable it either.

**Comment by dnouri** [comment]
> Ha, looking at the objective docs (which are still in the old format 
> and will probably be rewritten soon), I now remember why I decided to 
> look for a different theme ;) 
> http://lasagne.readthedocs.org/en/latest/modules/objectives.html
> 
> With numpydoc it does indeed look much better than the alternatives, 
> though.
> 

This 'old format' was just weird though. The vanilla Sphinx parameter 
syntax renders much nicer.

**Comment by skaae** [comment]
The new Docs layout is nice :) 

Any reason the Embedding layer is not showing up?

**Comment by MartinThoma** [comment]
Side note: The reference on http://lasagne.readthedocs.org/en/latest/modules/layers.html#lasagne.layers.dropout seems to be wrong:

> Dropout layer [R6],[R7]_.

The reference [R7] does not have a link.

**Comment by f0k** [comment]
> The reference [R7] does not have a link.

Ah, that's because of the empty line in between the two references in the `References` section. We need to edit that file quite a bit anyway -- both to split up the `layers` documentation and to document the shortcuts a bit more sensically than by duplicating the full class docstring. (I've shown how in the `nonlinearities` module.)

**Comment by benanne** [comment]
> Any reason the Embedding layer is not showing up? 

Because it's not listed in `docs/modules/layers.rst` yet, I guess.

It's unfortunate that we have to do this, because it's often forgotten. It would be nicer if Sphinx were able to discover all the `Layer` classes by itself. If anyone knows how to achieve it without exposing the submodules into which `lasagne.layers` is divided, please do tell :)

**Comment by skaae** [comment]
I have a suggestion for making the layer docs read more easily. 

We should create a short description for each group of layers, like the one for updates.py.

We were playing around with docs for the recurrent layers and were able to insert the descriptions with

```
Layer classes: recurrent layers
-------------------------------

.. automodule:: lasagne.layers.recurrent


.. autoclass:: CustomRecurrentLayer
    :members:

.. autofunction:: RecurrentLayer
    :members:

.. autoclass:: LSTMLayer
    :members:
```

Which might not be the correct way?

**Comment by f0k** [comment]
As Sander's TODO list says, we'd like to split up `layers.rst` into multiple files, one per submodule. `layers.rst` would just have overview tables similar to `nonlinearities.rst` then.

> We were playing around with docs for the recurrent layers and were able to insert the descriptions with [...]

Yes, that's the way we're currently doing it for all the other layer classes in [`layers.rst`](https://raw.githubusercontent.com/Lasagne/Lasagne/master/docs/modules/layers.rst) as well! And that's what would be moved to a sub-page.

**Comment by skaae** [comment]
I dont see any layers having descriptions? Maybe its not clear what I meant. 

In the screen shot if have clicked on recurrent layers and a short description of all recurrent layers is then shown.
![aaa](https://cloud.githubusercontent.com/assets/2623134/7899169/4af4ba4a-0719-11e5-8267-b84c2df4ae33.jpg)

I think it would be nice if all layer submodules had such a description.

**Comment by skaae** [comment]
I see the last todo now, so maybe that is 

`split lasagne.layers documentation into separate pages for each submodule`

:)

**Comment by f0k** [comment]
> I think it would be nice if all layer submodules had such a description. 

Ah, I see. Agreed, looks good, except that I'd also use `autosummary` tables instead of the list here :) This would need a module docstring for every submodule, and an `automodule` line in `layers.rst` (or the respective sub-page when we have them).

**Comment by skaae** [comment]
Do anyone know how to write latex equations that span several lines in the py file? 

The following equation is too long:
```
.. math:: \hat{h}^j_t = tanh(W\mathbf{x}_t + \mathbf{r}_t \odot (U\mathbf{h_{t-1}}))^j
```

**Comment by f0k** [comment]
> Do anyone know how to write latex equations that span several lines in the py file? 

Yes, with the `.. math::` notation, you can just wrap them. See "10. Notes" in the [numpydoc howto](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#sections). You may have to start them in the next line, like this:
```
.. math::
   foo bar foo
   bar foo bar
```
At least that's how I did it for the `lasagne.init` docstrings: https://github.com/Lasagne/Lasagne/blob/70f3da80c377fbc35e29edd3628e2e91529212bb/lasagne/init.py#L140-142

**Comment by skaae** [comment]
Thanks. I figured out that you need to end the .. math:: envirionment with a dobule line break for it to work.

**Comment by skaae** [comment]
I think MaxPool2DDNNLayer is missing in the docs? Is this the right place to report such things?

**Comment by benanne** [comment]
It might get lost in a long thread like this one. Maybe open a separate issue for it.

**Comment by benanne** [comment]
We're almost done :) Do we need something fancier for `regularization` than just the function docstrings? Personally I think what we have is fine for now, so then we just have to split up the layer docs into separate pages to be done with this monster of an issue.

**Comment by f0k** [comment]
> Do we need something fancier for `regularization` than just the function docstrings?

Yes, I think we do. It doesn't have a single example now. A short introduction, two tables and an example would be nice, similar to [objectives](http://lasagne.readthedocs.org/en/latest/modules/objectives.html).

> to be done with this monster of an issue.

It's a small and cute monster compared to #104 and #17 :)

**Comment by benanne** [comment]
Is anyone up for fleshing out the `regularization` page in the docs? That's pretty much the last task of any substance that still needs to be done, splitting up the layer docs into separate files should be quick enough that I can take care of it myself :)

**Comment by f0k** [comment]
Could do it, but I'll handle the MNIST example and the README first. Wouldn't be mad if anyone else did the `regularization` docstring!

**Comment by skaae** [comment]
Is the regularization docstring something similar to the overview in updates.py? 

Then i can give it a shot.

**Comment by benanne** [comment]
Yeah, pretty much :)

**Comment by skaae** [comment]
I created a PR for regularization docstring in https://github.com/Lasagne/Lasagne/pull/314

**Comment by f0k** [comment]
> splitting up the layer docs into separate files should be quick

Unfortunately, we cannot easily split it up because of the following Sphinx bug: https://github.com/sphinx-doc/sphinx/issues/1822
In short, this bug prevents the following to work:
```
Helper functions
----------------

.. autosummary::
    :nosignatures:
    
    get_output
    get_output_shape
```
It will skip generating entries for those because they are not actually defined in `lasagne.layers`, but in `lasagne.layers.helper`.  To fix that, we'd either need to use an older version of Sphinx (e.g., 1.2.3), or use the git version. I'm not sure if and how that is possible with readthedocs.

The alternative is to not use `autosummary` at all, but either define tables manually or just create bullet lists without descriptions for now.

**Comment by benanne** [comment]
Maybe let's postpone the splitting for now in that case. For the actual rendered docs it won't make much of a difference anyway.

**Comment by dnouri** [comment]
> To fix that, we'd either need to use an older version of Sphinx (e.g., 1.2.3), or use the git version. I'm not sure if and how that is possible with readthedocs.

I think we configured readthedocs to use `requirements-dev.txt`, which lists Sphinx 1.2.3 as a dependency: https://github.com/Lasagne/Lasagne/blob/master/requirements-dev.txt#L9  (This would also be the hook to use a Sphinx version from Git, if desired.)

**Comment by f0k** [comment]
Last checkpoint on the TODO list is done! Thanks to everybody who contributed!

**Comment by f0k** [comment]
Funny, how did `lasagne.layers.normalization` dodge under our radar? It wasn't in the TODO list and it wasn't in the docs either. It's not in numpydoc format either. I'll cope with that while adding the `BatchNormLayer`.

**Comment by benanne** [comment]
I guess nobody double-checked whether my list was complete :)
